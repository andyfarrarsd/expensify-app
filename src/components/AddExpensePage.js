import React from 'react';
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses'

export class AddExpensePage extends React.Component {
  
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/');
  };

  render() {
    return (
    <div>
      <h1>Add Expense</h1>
      <ExpenseForm
        onSubmit={this.onSubmit}
      />
    </div>  
  );
 };
}

// instead of a return implicitly return a value in this case an object
const mapDispatchToProps = (dispatch) => ({   
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
