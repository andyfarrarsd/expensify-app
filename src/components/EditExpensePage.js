import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { removeExpense, editExpense } from '../actions/expenses';


export class EditExpensePage extends React.Component {

  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  onRemove = () => {
    this.props.removeExpense( { id: this.props.expense.id });
    this.props.history.push('/');
  };

  render() {
    return (
    <div>
      <ExpenseForm
        expense={this.props.expense}
        onSubmit={this.onSubmit}
      />
      <button onClick={this.onRemove}> Remove </button>
    </div>
    );
  }
};

// instead of a return implicitly return a value in this case an object
const mapDispatchToProps = (dispatch, props) => ({   
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: (id) => dispatch(removeExpense(id))
});

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => {
      return expense.id === props.match.params.id;
    })
  };
};

//export default connect(mapStateToProps)(EditExpensePage);
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
