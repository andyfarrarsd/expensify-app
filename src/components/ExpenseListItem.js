// Export a stateless functional component
// Description, amount, createdAt
import React from 'react';
import { Link } from 'react-router-dom'
import moment from 'moment';
import numeral from 'numeral';

export const ExpenseListItem = ({ id, description, amount, createdAt }) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p> 
         {numeral(amount / 100).format('$0,0.00')}
         -
         { moment(createdAt).format('MMMM Do, YYYY') }
        </p>
    </div>
);

//Need to connect this component to the store to get access to props
// (state)(component to connent to state)
export default ExpenseListItem;
