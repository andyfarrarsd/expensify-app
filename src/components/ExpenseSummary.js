import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import getExpensesTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';


// Add total as an H1 block
export const ExpenseSummary = ( { expenseCount=0, expensesTotal=0} ) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
    const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00')

    return (
        <div>
           <h1>Viewing {expenseCount} {expenseWord} totaling { formattedExpensesTotal } </h1>
        </div>
    );
};

// This function is used to set up props based on current state. We set up the visibileExpenses array for this purpose
const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);

    return{
        expenseCount: visibleExpenses.length,
        expensesTotal: getExpensesTotal(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpenseSummary);
