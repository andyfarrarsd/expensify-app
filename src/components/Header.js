import React from 'react';
import { NavLink} from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';


// Add Header and basic links
export const Header = ( { startLogout }) => (
    <header>  
    <h1>Expensify</h1>
            <NavLink to="/" activeClassName="is-active" exact={true}>Dashboard </NavLink>
            <NavLink to="/create" activeClassName="is-active">Create Expense </NavLink>
            <button onClick={ startLogout}>Logout</button>
    </header>
);

// This is the wierd redux connection for action dispatching
const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});


export default connect(undefined, mapDispatchToProps)(Header);