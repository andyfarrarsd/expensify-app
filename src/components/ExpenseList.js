import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

 // Use array map to iterate through the expenses array creating an array of expenselistitems
export const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        {
            props.expenses.map((expense, index) => {
                return <ExpenseListItem key={expense.id} {...expense} /> // use spread operator to send all the elements of expense
            // Could also pass expense and then access props.each element in ELI
            })
        }
    </div>
);

// This function maps state to local props. below sets a props var that includes all our expenses
const mapStateToProps = (state) => {
    return{
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);