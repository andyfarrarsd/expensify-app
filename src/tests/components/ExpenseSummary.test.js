import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseSummary } from '../../components/ExpenseSummary';

test('should render expense total 1 item correctly', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={1} ExpenseTotal={1234} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render expense total multiple items correctly', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={3} ExpenseTotal={253500}/>);
    expect(wrapper).toMatchSnapshot();
});
