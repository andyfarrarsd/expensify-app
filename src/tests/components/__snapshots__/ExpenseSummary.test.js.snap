// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense total 1 item correctly 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totaling 
    $0.00
     
  </h1>
</div>
`;

exports[`should render expense total multiple items correctly 1`] = `
<div>
  <h1>
    Viewing 
    3
     
    expenses
     totaling 
    $0.00
     
  </h1>
</div>
`;
